### ISSUE TITLE
After upgrade to Google.Cloud.Logging.V2 (3.0.0) container restarts on LoggingServiceV2Client.WriteLogEntriesAsync call

### ISSUE THREAD

**Issue by slim1305** [googleapis/google-cloud-dotnet#4780]
- OS: docker image on mcr.microsoft.com/dotnet/core/aspnet:3.1-alpine + glibc package
  - .NET version: 3.1.3
  - Package name and version: Google.Cloud.Logging.V2 (3.0.0)
 - GKE cluster + gitlab + helm charts

Everything works fine if I downgrade to 2.X
#### Steps to reproduce

```
var client = LoggingServiceV2Client.Create(); //success
var platform = Platform.Instance(); //success
var resource = platform.Type == PlatformType.Unknown
                ? MonitoredResourceBuilder.GlobalResource
                : MonitoredResourceBuilder.FromPlatform(platform);
// success: platform.Type is "gke"
var response = await client.WriteLogEntriesAsync(
                    CreateLogName("development-platform-221211", "Probe2-WebApi-Dev"),
                    resource,
                    null,
                    new[]{ new LogEntry
                    {
                        Severity = LogSeverity.Error,
                        Timestamp = Timestamp.FromDateTimeOffset(DateTimeOffset.UtcNow),
                        TextPayload = "Test, Test, Test",
                    }});
```

On the last call container is restarting.
I use google cloud logging for a long time (1.5 years). But currently I need to implement gRPC service and have to upgrade to Google.Cloud.Logging.V2 (3.0.0).
May be I do something wrong could you please help.

Thanks!

**Comment by jskeet** [comment]
I strongly suspect that this is a duplicate of https://github.com/googleapis/google-cloud-dotnet/issues/3687 - please could you try the workaround in the [Stack Overflow question](https://stackoverflow.com/a/58708290/22656) to see if that fixes the problem?

**Comment by slim1305** [comment]
> I strongly suspect that this is a duplicate of #3687 - please could you try the workaround in the [Stack Overflow question](https://stackoverflow.com/a/58708290/22656) to see if that fixes the problem?

No, because I've already use glibc in my image like follow:
`
# Resolving gRPC dependencies in Alpine OS	
RUN curl -L -o /etc/apk/keys/sgerrand.rsa.pub https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub && \
    curl -L -O https://github.com/sgerrand/alpine-pkg-glibc/releases/download/$GLIBC_VERSION/glibc-$GLIBC_VERSION.apk && \
    apk add glibc-$GLIBC_VERSION.apk && \
    rm glibc-$GLIBC_VERSION.apk
`
Without glibc i've got exception about missed lib <don't remembe>.so. 
On alpine image I've got missed lib exception event on the 2.X version.

Any other ideas ?

**Comment by slim1305** [comment]
I'm going to provide tomorrow more info like:
- docker image code 
- sample

If it necessary ?

**Comment by jskeet** [comment]
A sample reproducing the problem would be *very* useful, yes.

(As an aside, please use triple backticks rather than single ones when posting multi-line code etc - otherwise it all comes out as a single line. I've edited both of the current comments, but it would be great not to have to do so in future :)

**Comment by slim1305** [comment]
Hi, 
In our application a lot of services and base docker image for all of them is:
```
FROM docker:17.12.0-ce as static-docker-source

FROM alpine:latest AS base
ENV GLIBC_VERSION=2.31-r0

COPY --from=static-docker-source /usr/local/bin/docker /usr/local/bin/docker

RUN apk add --no-cache \
    ca-certificates \
	curl \
	bash \
    \
    # .NET Core dependencies
    krb5-libs \
    libgcc \
    libintl \
    libssl1.1 \
    libstdc++ \
    zlib \
	icu

# Configure web servers to bind to port 80 when present
ENV ASPNETCORE_URLS=http://+:80 \
    # Enable detection of running in a container
    DOTNET_RUNNING_IN_CONTAINER=true
	
# Resolving gRPC dependencies in Alpine OS	
RUN curl -L -o /etc/apk/keys/sgerrand.rsa.pub https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub && \
    curl -L -O https://github.com/sgerrand/alpine-pkg-glibc/releases/download/$GLIBC_VERSION/glibc-$GLIBC_VERSION.apk && \
    apk add glibc-$GLIBC_VERSION.apk && \
    rm glibc-$GLIBC_VERSION.apk
```

Service docker image is:
```
FROM <link to the prev. one> AS base

EXPOSE 80
EXPOSE 443

ARG projectName
ENV dockerProjectName="./$projectName"
# Copy 
WORKDIR /app
COPY ./app .

ENTRYPOINT ["/bin/bash", "-c", "$dockerProjectName"]
```

Part of gitlab ci script to build app:
```
dotnet restore "${MAIN_PROJECT_PATH}/${PROJECT_FILE_NAME}" --packages $NUGET_PACKAGES_DIRECTORY -r alpine-x64 --source $NUGET_FEED --source $MYGET_FEED
dotnet publish "${MAIN_PROJECT_PATH}/${PROJECT_FILE_NAME}" -c Release -r alpine-x64 --self-contained true --version-suffix "${CI_PIPELINE_IID}" -o $APP_PATH --no-restore
```

Cluster is:
Master version: 1.15.9-gke.26
Master zone: europe-west2-b

On development env I use:
Image type: Container-Optimized OS (cos)
Machine type: custom-1-6656
Boot disk type: SSD persistent disk
Boot disk size (per node): 50 GB
Boot disk encryption: Google-managed
Preemptible nodes: Enabled

**Comment by jskeet** [comment]
A couple of things to suggest when tracking this down:

**Console app**

Try the same environment, but with a console app. Here's a simple app that just lists the log names - and only performs one RPC, to avoid creating any problems that way:

```csharp
using Google.Api.Gax.ResourceNames;
using Google.Cloud.Logging.V2;
using System;
using System.Linq;

namespace Issue4780
{
    class Program
    {
        static void Main(string[] args)
        {
            if (args.Length == 0)
            {
                Console.WriteLine("Please specify the project ID");
                return;
            }
            var client = LoggingServiceV2Client.Create();
            ProjectName projectName = ProjectName.FromProject(args[0]);
            var response = client.ListLogs(projectName).AsRawResponses().First();
            Console.WriteLine("First response");
            Console.WriteLine(response);
        }
    }
}
```

The only dependency needed for that is Google.Cloud.Logging.V2.

It would be great to know if that console app, when deployed in your environment, causes the same problem.

**Test locally**

It would be great to remove the GKE part from the problem as well. If the above fails in your GKE environment, does it *also* fail when running locally in Docker?

**Reduce to a single Dockerfile**

If it's still failing, can you reduce your series of Dockerfiles to a single file?

Basically I'm trying to get to the stage where we can reproduce this as simply as possible. I'm hoping that will only need three files: the C# source, the C# project, and a single Dockerfile.

Are you able to give all that a try? I'll see whether I can reproduce locally with the console app and the Dockerfile you've already shown, but I wouldn't be surprised if that worked... in which case I'm afraid some of the diagnostic steps will have to be done in your environment.

**Comment by slim1305** [comment]
Ok, but what about auth in local test ?
GKE node by default don't need auth json file or something else.
Should I localy use any auth with <service-account>.json or gcloud auth activate-service-account ?

Thanks,

**Comment by jskeet** [comment]
> Ok, but what about auth in local test ?

You can copy a service account JSON file into the container and set the GOOGLE_APPLICATION_CREDENTIALS environment variable to say where it is. That's probably the simplest approach.
I'm testing my app with your Dockerfile now... will post the results shortly.

**Comment by jskeet** [comment]
Aha - I've now got a segfault too. This is progress. You can probably leave it with me to investigate for the moment, unless you have any other suggestions.

**Comment by jskeet** [comment]
Okay, it looks like the fix in the linked Stack Overflow question doesn't work for Alpine 3.10 or 3.11. This is a [known gRPC issue](https://github.com/grpc/grpc/issues/21446). I've tried the workaround there and it works for my simplified Dockerfile:

```text
FROM mcr.microsoft.com/dotnet/core/runtime-deps:3.1-alpine3.11

RUN echo 'http://dl-cdn.alpinelinux.org/alpine/v3.8/main' >> /etc/apk/repositories && \
    apk update --no-cache && \
    apk add --no-cache bash libc6-compat=1.1.19-r11

# Publish first:
# dotnet publish -c Release -r alpine-x64 --self-contained true -o $PWD/app
WORKDIR /app
COPY ./app .
COPY ./credentials.json .
ENV GOOGLE_APPLICATION_CREDENTIALS=/app/credentials.json

ENTRYPOINT ["/bin/sh"]
```

Now trying with your more complex one...

**Comment by jskeet** [comment]
Success! Replacing your section starting `# Resolving gRPC dependencies` with

```text
RUN echo 'http://dl-cdn.alpinelinux.org/alpine/v3.8/main' >> /etc/apk/repositories && \
    apk update --no-cache && \
    apk add --no-cache bash libc6-compat=1.1.19-r11
```

worked for my console app. Now, I can't tell if there are any other issues that might cause within ASP.NET Core, unfortunately. This is outside the areas I know in depth - but I hope it's helpful.
If it doesn't work for you, it's probably worth posting more details on the [gRPC issue](https://github.com/grpc/grpc/issues/21446) where there are rather more knowledgeable users.

**Comment by slim1305** [comment]
Yee, your solution works fine. 
Great job, thanks.
![Screen](https://user-images.githubusercontent.com/17482095/78346687-0bbb4800-75a8-11ea-985a-0f668238ea0b.PNG)
